### PR TITLE
Nodejs week3

### PR DESCRIPTION
**Pull request by oyelajakenny**
Meal sharing endpoints
Added more parameters to the meal route
Created the meal endpoint and implemented all routes and parameters

### CODE REVIEW

File: api/src/routers/meals.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+mealsRouter.get("/", async (req, res) => {
+  const meals = await knex("meal").orderBy("id");
+  res.json(meals);
+});
+
+mealsRouter.post("/", async (req, res, next) => {
+  try {
+    console.log(req.body);
+    const data = req.body;
+    await knex("Meal").insert(data);
```
Comment (kralle333) [inline api/src/routers/meals.js:15]: I assume its `"meal"` instead of `"Meal"` :)

File: api/src/routers/meals.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  try {
+    const id = req.params.id;
+    const meals = await knex("meal").where({ id: id }).first();
+    if (!meals) {
+      res.status(404).json({ error: "No meal found" });
+    } else {
+      res.json(meals);
+    }
+    res.status(200).json({
+      success: true,
```
Comment (kralle333) [inline api/src/routers/meals.js:32]: getting an error code 200 should be good enough, no need to send an additional boolean here.

File: api/src/routers/meals.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    next(res.status(500).json({ error: "Unable to retrieve meal" }));
+  }
+});
+
+mealsRouter.put("/:id", async (req, res) => {
+  try {
+    const id = req.params.id;
+    const data = req.body;
+    await knex("meal").where({ id: id }).update(data);
+    if (!id) {
```
Comment (kralle333) [inline api/src/routers/meals.js:45]: I guess we can check this before querying the database :)

File: api/src/routers/meals.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    res.status(500).json({ error: "Unable to update meal" });
+  }
+});
+
+mealsRouter.delete("/:id", async (req, res, next) => {
+  try {
+    const id = req.params.id;
+    await knex("meal").where({ id: id }).del();
+    if (!id) {
+      res.status(404).json({ error: "No meal with this ID found" });
```
Comment (kralle333) [inline api/src/routers/meals.js:60]: are there no meals with this id, or is the request simply invalid? Tip: use the value returned from the delete statement to return the proper response.

File: api/src/routers/reservations.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+const reservationsRouter = express.Router();
+
+reservationsRouter.get("/", async (req, res) => {
+  const reservations = await knex("reservation").orderBy("id");
+  res.json(reservations);
+});
+
+reservationsRouter.post("/", async (req, res, next) => {
+  try {
+    console.log(req.body);
```
Comment (kralle333) [inline api/src/routers/reservations.js:13]: I guess useful during testing, but for now we don't need this logging :)

File: api/src/routers/reservations.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  } catch (error) {
+    res.status(500).json({ error: "Unable to update reservation" });
+  }
+});
+
+reservationsRouter.delete("/:id", async (req, res) => {
+  try {
+    const id = req.params.id;
+    await knex("Reservation").where({ id: id }).del();
+    res.status(200).json({ message: "deleted successfully" });
```
Comment (kralle333) [inline api/src/routers/reservations.js:51]: what if there is no reservation with that id? Similar solution you can use in the meals routes

File: api/src/routers/reviews.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  try {
+    const reviews = await knex("review").where("meal_id", meal_id).select("*");
+
+    if (reviews.length === 0) {
+      res.status(404).json({ error: "No review found" });
+    }
+    res.json(reviews);
+  } catch (error) {
+    console.log(error);
+    res.status(505).json({ error: "Internal server error" });
```
Comment (kralle333) [inline api/src/routers/reviews.js:26]: 500 should suffice in this situation, bonus if you remove the console.log ;)

File: api/src/routers/reviews.js [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+  }
+});
+
+reviewRouter.get("/:meal_id/reviews", async (req, res) => {
+  const { meal_id } = req.params;
+  try {
+    const reviews = await knex("review").where("meal_id", meal_id).select("*");
+
+    if (reviews.length === 0) {
+      res.status(404).json({ error: "No review found" });
```
Comment (kralle333) [inline api/src/routers/reviews.js:21]: Here the proper response is actually 200! Weird right, but 404 is reserved for cases where the user expected some specific id to exist but didnt. Here the users asks for all the reviews for a specific meal and gets 0 if there are none, so it works as expected.

File: api/src/routers/reviews.js [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+  try {
+    const review = await knex("review").where({ id }).first();
+    if (!review) {
+      res.status(404).json({ error: "No review found" });
+    } else {
+      res.status(200).json(review);
+    }
+  } catch (error) {
+    console.log(error);
+    res.status(505).json({ error: "Internal server error" });
```
Comment (kralle333) [inline api/src/routers/reviews.js:51]: 500 here too ;)

File: api/src/routers/reviews.js [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+  } catch (error) {
+    console.log(error);
+    res.status(500).json({ error: "Unable to update review" });
+  }
+});
+
+reviewRouter.delete("/:id", async (req, res) => {
+  const { id } = req.params;
+  try {
+    await knex("review").where({ id }).del();
```
Comment (kralle333) [inline api/src/routers/reviews.js:70]: same comment as in the other delete routes

**Review by kralle333 — CHANGES_REQUESTED** [review]
Great work! Overall your code is very solid.

There are some small problems here and there I would like you to fix.

Keep it up!

File: api/src/routers/meals.js [hunk 11]
```diff
@@ -0,0 +1,69 @@
+import express from "express";
+import knex from "../database_client.js";
+
+const mealsRouter = express.Router();
+
+mealsRouter.get("/", async (req, res) => {
```
Comment (kralle333) [inline api/src/routers/meals.js:6]: I believe you are missing one of the tasks here?
image

File: api/src/routers/meals.js [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    });
+  } catch (error) {
+    next(res.status(500).json({ error: "Unable to retrieve meal" }));
+  }
+});
+
+mealsRouter.put("/:id", async (req, res) => {
+  try {
+    const id = req.params.id;
+    const data = req.body;
```
Comment (kralle333) [inline api/src/routers/meals.js:43]: missing validation of the request here

File: api/src/routers/reservations.js [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    }
+  } catch (error) {
+    res.status(500).json({ error: "Unable to get reservation" });
+  }
+});
+
+reservationsRouter.put("/:id", async (req, res) => {
+  try {
+    const id = req.params.id;
+    const data = req.body;
```
Comment (kralle333) [inline api/src/routers/reservations.js:39]: missing validation of the request here

File: api/src/routers/reviews.js [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+    res.json(reviews);
+  } catch (error) {
+    console.log(error);
+    res.status(505).json({ error: "Internal server error" });
+  }
+});
+
+reviewRouter.post("/", async (req, res, next) => {
+  try {
+    const data = req.body;
```
Comment (kralle333) [inline api/src/routers/reviews.js:32]: missing validation of the request here

File: api/src/routers/reviews.js [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+  } catch (error) {
+    console.log(error);
+    res.status(505).json({ error: "Internal server error" });
+  }
+});
+
+reviewRouter.put("/:id", async (req, res) => {
+  const { id } = req.params;
+  try {
+    const data = req.body;
```
Comment (kralle333) [inline api/src/routers/reviews.js:58]: missing validation of the request here

File: api/src/routers/meals.js (hunk 11 above)
Comment (oyelajakenny) [reply api/src/routers/meals.js:6]: Hi @kralle333, I attempted all tasks, I have now added comments to distinguish them.

**Comment by oyelajakenny** [comment]
Hi @kralle333 , apologies for the late feedback. I have now implemented all recommendations

**Pull request merged**